File: glibc/nptl/test-rwlockattr-printers.py
# Common tests for the RWLockPrinter and RWLockAttributesPrinter classes.
#
# Copyright (C) 2016-2021 Free Software Foundation, Inc.
# This file is part of the GNU C Library.
#
# The GNU C Library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License as published by the Free Software Foundation; either
# version 2.1 of the License, or (at your option) any later version.
#
# The GNU C Library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public
# License along with the GNU C Library; if not, see
# <https://www.gnu.org/licenses/>.

import sys

from test_printers_common import *

test_source = sys.argv[1]
test_bin = sys.argv[2]
printer_files = sys.argv[3:]
printer_names = ['global glibc-pthread-locks']

try:
    init_test(test_bin, printer_files, printer_names)
    go_to_main()

    check_debug_symbol('struct pthread_rwlockattr')

    rwlock_var = 'rwlock'
    rwlock_to_string = 'pthread_rwlock_t'

    attr_var = 'attr'
    attr_to_string = 'pthread_rwlockattr_t'

    break_at(test_source, 'Set kind')
    continue_cmd() # Go to test_setkind_np
    next_cmd(2)
    test_printer(rwlock_var, rwlock_to_string, {'Prefers': 'Readers'})
    test_printer(attr_var, attr_to_string, {'Prefers': 'Readers'})
    next_cmd(2)
    test_printer(rwlock_var, rwlock_to_string, {'Prefers': 'Writers'})
    test_printer(attr_var, attr_to_string, {'Prefers': 'Writers'})
    next_cmd(2)
    test_printer(rwlock_var, rwlock_to_string, {'Prefers': 'Writers no recursive readers'})
    test_printer(attr_var, attr_to_string, {'Prefers': 'Writers no recursive readers'})

    break_at(test_source, 'Set shared')
    continue_cmd() # Go to test_setpshared
    next_cmd(2)
    test_printer(rwlock_var, rwlock_to_string, {'Shared': 'Yes'})
    test_printer(attr_var, attr_to_string, {'Shared': 'Yes'})
    next_cmd(2)
    test_printer(rwlock_var, rwlock_to_string, {'Shared': 'No'})
    test_printer(attr_var, attr_to_string, {'Shared': 'No'})

    continue_cmd() # Exit

except (NoLineError, pexpect.TIMEOUT) as exception:
    print('Error: {0}'.format(exception))
    result = FAIL

except DebugError as exception:
    print(exception)
    result = UNSUPPORTED

else:
    print('Test succeeded.')
    result = PASS

exit(result)

File: glibc/nptl/test-rwlockattr-printers.c
/* Helper program for testing the pthread_rwlock_t and pthread_rwlockattr_t
   pretty printers.

   Copyright (C) 2016-2021 Free Software Foundation, Inc.
   This file is part of the GNU C Library.

   The GNU C Library is free software; you can redistribute it and/or
   modify it under the terms of the GNU Lesser General Public
   License as published by the Free Software Foundation; either
   version 2.1 of the License, or (at your option) any later version.

   The GNU C Library is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
   Lesser General Public License for more details.

   You should have received a copy of the GNU Lesser General Public
   License along with the GNU C Library; if not, see
   <https://www.gnu.org/licenses/>.  */

/* Keep the calls to the pthread_* functions on separate lines to make it easy
   to advance through the program using the gdb 'next' command.  */

#include <pthread.h>

#define PASS 0
#define FAIL 1

/* Need these so we don't have lines longer than 79 chars.  */
#define SET_KIND(attr, kind) pthread_rwlockattr_setkind_np (attr, kind)
#define SET_SHARED(attr, shared) pthread_rwlockattr_setpshared (attr, shared)

static int rwlock_reinit (pthread_rwlock_t *rwlock,
			  const pthread_rwlockattr_t *attr);
static int test_setkind_np (pthread_rwlock_t *rwlock,
			    pthread_rwlockattr_t *attr);
static int test_setpshared (pthread_rwlock_t *rwlock,
			    pthread_rwlockattr_t *attr);

int
main (void)
{
  pthread_rwlock_t rwlock;
  pthread_rwlockattr_t attr;
  int result = FAIL;

  if (pthread_rwlockattr_init (&attr) == 0
      && pthread_rwlock_init (&rwlock, NULL) == 0
      && test_setkind_np (&rwlock, &attr) == PASS
      && test_setpshared (&rwlock, &attr) == PASS)
    result = PASS;
  /* Else, one of the pthread_rwlock* functions failed.  */

  return result;
}

/* Destroys RWLOCK and re-initializes it using ATTR.  */
static int
rwlock_reinit (pthread_rwlock_t *rwlock, const pthread_rwlockattr_t *attr)
{
  int result = FAIL;

  if (pthread_rwlock_destroy (rwlock) == 0
      && pthread_rwlock_init (rwlock, attr) == 0)
    result = PASS;

  return result;
}

/* Tests setting whether the rwlock prefers readers or writers.  */
static int
test_setkind_np (pthread_rwlock_t *rwlock, pthread_rwlockattr_t *attr)
{
  int result = FAIL;

  if (SET_KIND (attr, PTHREAD_RWLOCK_PREFER_READER_NP) == 0 /* Set kind.  */
      && rwlock_reinit (rwlock, attr) == PASS
      && SET_KIND (attr, PTHREAD_RWLOCK_PREFER_WRITER_NP) == 0
      && rwlock_reinit (rwlock, attr) == PASS
      && SET_KIND (attr, PTHREAD_RWLOCK_PREFER_WRITER_NONRECURSIVE_NP) == 0
      && rwlock_reinit (rwlock, attr) == PASS)
    result = PASS;

  return result;
}

/* Tests setting whether the rwlock can be shared between processes.  */
static int
test_setpshared (pthread_rwlock_t *rwlock, pthread_rwlockattr_t *attr)
{
  int result = FAIL;

  if (SET_SHARED (attr, PTHREAD_PROCESS_SHARED) == 0 /* Set shared.  */
      && rwlock_reinit (rwlock, attr) == PASS
      && SET_SHARED (attr, PTHREAD_PROCESS_PRIVATE) == 0
      && rwlock_reinit (rwlock, attr) == PASS)
    result = PASS;

  return result;
}
